src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -427,15 +427,15 @@ private BoundLambda ReallyBind(NamedTypeSymbol delegateType)
             // when binding for real (not for return inference), there is still
             // a good chance that we could reuse a body of a lambda previously bound for 
             // return type inference.
-            var cacheKey = ReturnInferenceCacheKey.Create(delegateType, IsAsync);
+            var cacheKey = ReturnInferenceCacheKey.Create(binder, delegateType, IsAsync);
 
             BoundLambda returnInferenceLambda;
-            if (_returnInferenceCache.TryGetValue(paramInfo, out returnInferenceLambda) && returnInferenceLambda.InferredFromSingleType)
+            if (_returnInferenceCache.TryGetValue(cacheKey, out returnInferenceLambda) && returnInferenceLambda.InferredFromSingleType)
             {
                 lambdaSymbol = returnInferenceLambda.Symbol;
-                var lambdaRetType = lambdaSymbol.ReturnType.TypeSymbol;
-                if ((object)LambdaSymbol.InferenceFailureReturnType != lambdaRetType &&
-                    lambdaRetType == returnType && lambdaSymbol.RefKind == refKind)
+                var lambdaReturnType = lambdaSymbol.ReturnType;
+                if ((object)LambdaSymbol.InferenceFailureReturnType != lambdaReturnType &&
+                    lambdaReturnType.Equals(returnType, TypeCompareKind.CompareNullableModifiersForReferenceTypes) && lambdaSymbol.RefKind == refKind)
```

Comment:
I don't remember witch comparison is stronger, I believe we should use strongest flavor here.